libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -187,14 +171,157 @@ Algorithm selectAlgorithm(TangentSpaceMeshInput *input) {
     return outAlgo;
 }
 
-#undef IS_INPUT_TYPE
+// The paper uses a Z-up world basis, which has been converted to Y-up here
+void frisvadKernel(const float3& N, float3& T, float3& B) noexcept {
+    if (N.y < -1.0f + std::numeric_limits<float>::epsilon()) {
+        // Handle the singularity
+        T = float3{-1.0f, 0.0f, 0.0f};
+        B = float3{0.0f, 0.0f, -1.0f};
+        return;
+    }
+    const float a = 1.0f / (1.0f + N.y);
+    const float b = -N.z * N.x * a;
+    T = float3{b, -N.z, 1.0f - N.z * N.z * a};
+    B = float3{1.0f - N.x * N.x * a, -N.x, b};
+}
+
+void hughesMollerKernel(const float3& N, float3& T, float3& B) noexcept {
+    if (abs(N.x) > abs(N.z) + std::numeric_limits<float>::epsilon()) {
+        T = float3{-N.y, N.x, 0.0f};
+    } else {
+        T = float3{0.0f, -N.z, N.y};
+    }
+    T = normalize(T);
+    B = cross(N, T);
+}
+
+void normalsOnlyMethod(const TangentSpaceMeshInput* input, TangentSpaceMeshOutput* output,
+    NormalsOnlyKernelPtr kernel) noexcept {
+    const size_t vertexCount = input->vertexCount;
+    quatf* quats = new quatf[vertexCount];
+
+    const float3* normal = input->normals;
+    size_t nstride = input->normalStride ? input->normalStride : sizeof(float3);
+
+    for (size_t qindex = 0; qindex < vertexCount; ++qindex) {
+        float3 n = *normal;
+        float3 b, t;
+        kernel(n, t, b);
+        quats[qindex] = mat3f::packTangentFrame({t, b, n});
+        normal = (const float3*) (((const uint8_t*) normal) + nstride);
```

Comment:
I think you can use the `pointerAdd` here.

Reply:
done